extensions/amp-selector/0.2/amp-selector.js

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {CSS} from '../../../build/amp-selector-0.1.css';
+import {Option, Selector} from './selector';
+import {PreactBaseElement} from '../../../src/preact/base-element';
+import {
+  closestAncestorElementBySelector,
+  scopedQuerySelector,
+} from '../../../src/dom';
+import {dev, userAssert} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {isArray, toArray} from '../../../src/types';
+import {isExperimentOn} from '../../../src/experiments';
+
+/** @const {string} */
+const TAG = 'amp-selector';
+
+class AmpSelector extends PreactBaseElement {
+  /** @override */
+  init() {
+    if (!this.element.hasAttribute('role')) {
+      this.element.setAttribute('role', 'listbox');
+    }
+    if (this.element.hasAttribute('multiple')) {
+      this.element.setAttribute('aria-multiselectable', 'true');
+    }
+    if (this.element.hasAttribute('disabled')) {
+      this.element.setAttribute('aria-disabled', 'true');
+    }
+    const getOptionState = () => {
+      const children = [];
+      const optionChildren = toArray(this.element.querySelectorAll('[option]'));
+
+      const value = [];
+      optionChildren
+        // Skip options that are themselves within an option
+        .filter((child) => !this.getOptionElement_(child.parentElement))
+        .forEach((child) => {
+          const option = child.getAttribute('option');
+          if (!child.hasAttribute('role')) {
+            child.setAttribute('role', 'option');
+          }
+          const isDisabled = child.hasAttribute('disabled');
+          if (isDisabled) {
+            child.setAttribute('aria-disabled', 'true');
+          }
+          const props = {
+            option,
+            disabled: isDisabled,
+            type: 'Slot',
+            retarget: true,
+            assignedElements: [child],
+            postRender: () => {
+              // Skip mutations to avoid cycles.
+              mu.takeRecords();
+            },
+            getOption: (e) =>
```

Comment:
Unfortunately the `Slot` elements only mimic the structure of the options but are not directly attached to the options themselves. In actuality, the first `Slot` in the shadow contains all options:
<img width="527" alt="Screen Shot 2020-04-03 at 2 56 34 PM" src="https://user-images.githubusercontent.com/10456171/78395441-7cb82b00-75bb-11ea-976a-7938e368c631.png">

So using anything like `props.option` or `props.assignedElements[0]` directly will always refer to the first option in the list of options, regardless of which among the DOM nodes was actually interacted with.

There is the option to attach the slots explicitly with corresponding `names`, but this disregards non-option elements in the original DOM structure. e.g. The `<div>` in the center of the above is not itself an option but separates the images into two rows. 

We could attach all children nodes to slots but then something like this also becomes difficult to manage:
```html
<amp-selector>
   <div option="1"></div>
   <div>
      <div option="2"></div>
   </div>
   <div></div>
</amp-selector>
```

Reply:
It'd definitely be the case if we attach the actual shadow DOM. That's why I was proposing to use a template tag instead. That way the component structure will be disconnected from the actual DOM and we can use a `Slot` (or another special-need component) to bridge the attributes and events between the APIs.

Reply:
And to clarify, this doesn't have to be the `Slot` component. We could create a special-use `Option` component for bridging. But it would have some overlap with what we do in the `Slot`.